File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int rootIdx;
    public TreeNode bstFromPreorder(int[] preorder) {
        rootIdx = 0;
        return helper(preorder, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }
    private TreeNode helper(int[] array, int minValue, int maxValue) {
        if (rootIdx >= array.length || array[rootIdx] <= minValue ||
            array[rootIdx] >= maxValue) return null;
        
        TreeNode tree = new TreeNode(array[rootIdx]);
        rootIdx++;
        tree.left = helper(array, minValue, tree.val);
        tree.right = helper(array, tree.val, maxValue);
        return tree;
    }
}

File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def bstFromPreorder(self, preorder):
        """
        :type preorder: List[int]
        :rtype: TreeNode
        """
        global rootIdx
        rootIdx = 0
        return self.helper(preorder, -float("inf"), float("inf"))
    
    def helper(self, array, minValue, maxValue):
        global rootIdx
        if rootIdx >= len(array) or array[rootIdx] <= minValue \
            or array[rootIdx] >= maxValue: return None
            
        tree = TreeNode(array[rootIdx])
        rootIdx += 1
        tree.left = self.helper(array, minValue, tree.val)
        tree.right = self.helper(array, tree.val, maxValue)
        return tree